p2p-tests/lib/p2p.ts

```diff
@@ -104,66 +86,151 @@ export class LinePrefix extends stream.Transform {
-  ]);
+interface PeerParams {
+  namespace: string;
+  ipAddress: string;
+  startArgs: string[];
+  dataPath: string;
 }
 
-export function startPeer(
-  namespace: string,
-  ip: string,
-  args: string[]
-): ProxyClient.ProxyClient {
-  initPeer(namespace);
-
-  const childProcess = spawnInNamespace(
-    namespace,
-    [
-      path.join(BIN_PATH, "radicle-proxy"),
-      "--peer-listen",
-      `${ip}:8776`,
-      "--http-listen",
-      `${ip}:17246`,
-      "--skip-remote-helper-install",
-      "--unsafe-fast-keystore",
-      "--insecure-http-api",
-      "--dev-log",
-      ...args,
-    ],
-    {
-      RAD_HOME: workspacePath([namespace]),
+class Peer {
+  public dataPath: string;
+  public radHome: string;
+  public namespace: string;
+  public ipAddress: string;
+  public startArgs: string[];
+
+  _childProcess: execa.ExecaChildProcess | undefined = undefined;
+
+  constructor({ namespace, ipAddress, startArgs, dataPath }: PeerParams) {
+    this.dataPath = dataPath;
+    this.radHome = path.join(dataPath, `${namespace}-rad-home`);
+    this.namespace = namespace;
+    this.ipAddress = ipAddress;
+    this.startArgs = startArgs;
+
+    fs.mkdirsSync(this.radHome);
+  }
+
+  get childProcess() {
+    if (this._childProcess) {
+      return this._childProcess;
+    } else {
+      throw new Error(
+        "Tried to access childProcess before it was initialized."
+      );
     }
-  );
+  }
 
-  processes.push(childProcess);
+  async stop(): Promise<void> {
+    if (!this.childProcess) {
+      throw new Error("Tried to stop() process that wasn't started.");
+    }
+
+    this.childProcess.kill("SIGTERM");
+    await this.childProcess;
+  }
 
-  return new ProxyClient.ProxyClient(`http://${ip}:17246`);
+  async kill(): Promise<void> {
+    if (!this.childProcess) {
+      throw new Error("Tried to kill() process that wasn't started.");
+    }
+
+    try {
+      this.childProcess.kill("SIGKILL");
+      await this.childProcess;
+    } catch {
+      // Ignore errors.
+    }
+  }
 }
 
-export function startSeed(namespace: string, args: string[]): void {
-  const radHome = workspacePath([namespace]);
-  fs.removeSync(radHome);
-  fs.mkdirsSync(radHome);
+export class UpstreamPeer extends Peer {
+  _proxyClient: ProxyClient.ProxyClient | undefined = undefined;
+  public checkoutPath: string;
+  public passphrase: string;
 
-  const childProcess = spawnInNamespace(
-    namespace,
-    [path.join(BIN_PATH, "upstream-seed"), "--rad-home", radHome, ...args],
-    {}
-  );
+  constructor({ namespace, ipAddress, startArgs, dataPath }: PeerParams) {
+    super({ namespace, ipAddress, startArgs, dataPath });
+    this.checkoutPath = path.join();
+    this.passphrase = namespace;
+
+    this.init();
+  }
+
+  init(): void {
+    execa.sync(path.join(BIN_PATH, "radicle-proxy-init"), [
+      this.namespace,
+      "--key-passphrase",
+      this.passphrase,
+      "--rad-home",
+      this.radHome,
+    ]);
+  }
```

Comment:
Can we inline this? The fewer methods a class has the easier it is to understand. And in the case the logic is tiny.

Reply:
d94331be949f9ad2912a5d8bf91b4284c61c4e6d